source/main.c

```diff
@@ -36,108 +36,114 @@ static int get_auth(struct thread *td, struct get_auth_args *args) {
-			printf_notification("Cannot read dump-list.txt on the USB drive.");
-			return 0;
-		}
-
-		i += n;
-	}
-
-	buf[st.st_size] = 0;
-	close(fd);
-
-	// parse list file
-	next = buf;
-
-	for (;;) {
-		char *p;
-
-		// check if no more entries
-		if (!*next) {
-			break;
-		}
-
-		// get target file
-		if (!(p = index(next, ':'))) {
-			printf_notification("dump-list.txt on the USB drive has invalid format.");
-			return 0;
-		}
-
-		*p = 0;
-
-		// get auth info
-		req.file = next;
-
-		if (kexec(get_auth, &req) < 0) {
-			printf_notification("Failed to dump auth info for %s.", next);
-			return 0;
-		}
-
-		// get output file
-		next = p + 1;
-
-		if (p = index(next, '\n')) {
-			*p = 0;
-
-			if (p[-1] == '\r') {
-				p[-1] = 0;
-			}
-		}
-
-		// write to usb drive
-		fd = open(next, O_WRONLY | O_CREAT | O_TRUNC, 0777);
-
-		if (fd < 0) {
-			printf_notification("Cannot create %s.", next);
-			return 0;
-		}
-
-		write(fd, req.buf, 136);
-		close(fd);
-
-		// move to next line
-		if (!p) {
-			break;
-		}
-
-		next = p + 1;
-	}
-
-	free(buf);
-	printf_notification("Auth info dump completed!");
-
-	return 0;
+    int fd, write_result;
+    struct stat st;
+    char *buf, *next;
+    size_t i;
+    struct get_auth_req req;
+
+    initKernel();
+    initLibc();
+    jailbreak();
+    mmap_patch();
+    initSysUtil();
+
+    // open list file
+    fd = open("/mnt/usb0/dump-list.txt", O_RDONLY, 0);
+    if (fd < 0) {
+        printf_notification("Cannot open dump-list.txt on the USB drive.");
+        return 0;
+    }
+
+    if (fstat(fd, &st) < 0) {
+        printf_notification("Cannot determine the size of dump-list.txt on the USB drive.");
+        close(fd);
+        return 0;
+    }
+
+    // read list file
+    buf = malloc(st.st_size + 1);
+    if (!buf) {
+        printf_notification("Failed to allocate memory.");
+        close(fd);
+        return 0;
+    }
+
+    i = 0;
+    while (i < (size_t)st.st_size) {
+        ssize_t n = read(fd, buf + i, st.st_size - i);
+        if (n < 0) {
+            printf_notification("Cannot read dump-list.txt on the USB drive.");
+            free(buf);
+            close(fd);
+            return 0;
+        } else if (n == 0) {
+            break;
+        }
+        i += n;
+    }
+
+    buf[st.st_size] = '\0';
+    close(fd);
+
+    // parse list file
+    next = buf;
+    for (;;) {
+        char *p;
+
+        if (!*next) {
+            break;
+        }
+
+        p = strchr(next, ':');
+        if (!p) {
+            printf_notification("dump-list.txt on the USB drive has invalid format.");
+            free(buf);
+            return 0;
+        }
+        *p = '\0';
+
+        req.file = next;
+        if (kexec(get_auth, &req) < 0) {
+            printf_notification("Failed to dump auth info for\n%s", next);
+            free(buf);
+            return 0;
+        }
+
+        next = p + 1;
+        p = strchr(next, '\n');
+        if (p) {
+            if (p > next && p[-1] == '\r') {
```

Comment:
Why check `p > next`?

Reply:
Basically just a safety double check to ensure we don't read memory before the string starts by accident.

Reply:
Where did you get this code from?

Reply:
Eh, mainly just knowing how unsafe C code can be, I was just adding anything I could to make sure it didn't break lol.

Reply:
Yeah, I'm removing that check... That was kind of idiotic on my part XD... A double check made me kinda realize "Oh, wait, that won't happen."